packages/next-swc/crates/core/tests/fixture/server-actions/server/25/input.js

```diff
@@ -0,0 +1,34 @@
+import deleteFromDb from 'db'
+
+export function Item({ id1, id2 }) {
+  id1++
+  return (() => {
+    id1++
+    return <Button action={deleteItem}>Delete</Button>
+  })()
+
+  async function deleteItem() {
+    'use server'
+    await deleteFromDb(id1)
+    await deleteFromDb(id2)
+  }
+}
+
+// In this example, if Button immediately executes the action, different ids should
+// be passed.
```

Comment:
However if Button does _not_ immediately execute the action then the same (double-incremented) id should be passed for both, which doesn't seem to be handled correctly with this compilation.

In addition, I believe we don't even let you run server actions at render time so I'm not sure it makes sense to optimize for that case?

Reply:
Another option might be to forbid capturing any mutable bindings and add a constraint that any captured variables must be assigned before the closure is referenced? Best practice is to have no interior mutability anywhere within props and arguably a captured `let` that gets reassigned is a type of mutability.

But I would argue the current compilation here is incorrect.

Reply:
Yeah I later realized that we should just disallow mutating any captured variables, as it's impossible to pass any mutability across the boundary.